### PR TITLE
registry: Allow configuring the logger

### PR DESCRIPTION
**Pull request by nhooyr**

### CODE REVIEW

**Comment by jonjohnsonjr** [comment]
cc @clrprod

**Comment by clrprod** [comment]
Sorry for the delay taking a look now.

File: pkg/registry/registry.go [hunk 1]
```diff
@@ -54,18 +55,44 @@ func (v *v) v2(resp http.ResponseWriter, req *http.Request) *regError {
...
+	return NewWithOptions(nil)
+}
+
+// Options describes the available options
+// for creating the registry.
+type Options struct {
+	// Log is used to log requests.
+	// If nil, the global logger is used.
+	Log *log.Logger
+}
```
Comment (clrprod) [inline pkg/registry/registry.go]: Looks awesome - I think this is definitely a pretty reasonable thing to change and thank you for sending in the change.

They style in go is generally to define a option type - i.e.

type Option func(o *registry)

And then have New take variadic arguments - i.e.
New(o Option...)

and in the body do something like

opts := &registry{log:defaultLog}
for _, o := range options {
 o(opts)
}

And define the logger option like
func WithLogger(l log.Type) Option

so you can call New() to get default behaviour, New(WithLogger(log.SomethingComplicated)) to get more interesting behaviour.

Let me know if that makes sense - I'm happy to have the log be configurable, just want to make sure the option pattern is nice for the user. I'm also happy to make these changes if I'm asking for too much work from you.

Comment (nhooyr) [reply pkg/registry/registry.go]: > They style in go is generally to define a option type - i.e.

Its another pattern for sure but I wouldn't say its generally better. The stdlib for example doesn't use it at all and even though later packages did, the new go-cloud package made by the Go team explicitly decided against using them in favour of optional structs.

See https://github.com/google/go-cloud/issues/908#issuecomment-445034143

Comment (nhooyr) [reply pkg/registry/registry.go]: > I'm also happy to make these changes if I'm asking for too much work from you.

Feel free to take over.

Comment (jonjohnsonjr) [reply pkg/registry/registry.go]: I think both approaches are reasonable. There are definitely some hard to solve problems with the functional options approach, but it looks so much nicer in some cases 😄 

We should keep in mind any other extension points e.g. https://github.com/google/go-containerregistry/issues/488

Comment (clrprod) [reply pkg/registry/registry.go]: I pushed a version of this in https://github.com/google/go-containerregistry/pull/537 - let me know what you think :D

**Comment by nhooyr** [comment]
Closed in favour of #537

**Pull request closed without merging**